backend/src/books/books.service.ts

```diff
@@ -1,13 +1,60 @@
 import { Injectable, HttpService } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository } from 'typeorm';
+import * as Papa from 'papaparse';
+import { Readable } from 'stream';
+import { validate, ValidationError } from 'class-validator';
 
-import { BookQueryResponse } from './books';
-import { Book } from '../../../interfaces/books';
+import { BookQueryResponse, BookUploadResult } from './books';
+import { Book as BookType } from '../../../interfaces/books';
+import { Book } from './book.entity';
+import { BookDto } from './book.dto';
+
+/** https://stackoverflow.com/a/54136803 */
+const bufferToStream = (buffer: Buffer): Readable => {
+  const readableStream = new Readable({
+    read(): void {
+      this.push(buffer);
+      this.push(null);
+    },
+  });
+
+  return readableStream;
+};
+
+const validateBooks = async (
+  books: BookDto[],
+): Promise<{
+  books: BookDto[];
+  errors: Record<string, ValidationError[]>;
+}> => {
+  const allErrors: Record<string, ValidationError[]> = {};
+
+  const validatedBooks = (
+    await Promise.all(
+      books.map(async book => {
+        const errors = await validate(book);
+        if (errors.length > 0) {
+          allErrors[book.isbn] = errors;
+          return undefined;
+        } else {
+          return book;
+        }
+      }),
+    )
+  ).filter(book => book !== undefined);
+
+  return { books: validatedBooks, errors: allErrors };
+};
 
 @Injectable()
 export class BooksService {
-  constructor(private httpService: HttpService) {}
+  constructor(
+    private httpService: HttpService,
+    @InjectRepository(Book) private readonly booksRepository: Repository<Book>,
+  ) {}
 
-  async getBooks(query: string): Promise<Book[]> {
+  async getBooks(query: string): Promise<BookType[]> {
```

Comment:
It is just a personal opinion but I find it easier to deal with observables. On the other hand this is not really a stream of data per say. No action point is needed here though in case you would like to keep the promises! 😊